File: android/app/src/main/kotlin/com/negfeed/epyc/ui/screens/GuessViewModel.kt
package com.negfeed.epyc.ui.screens

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.negfeed.epyc.AppContainer
import com.negfeed.epyc.core.AtomAddress
import com.negfeed.epyc.core.DrawingEvent
import com.negfeed.epyc.core.GameAtomState
import com.negfeed.epyc.core.GameNavigation
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.launch

/**
 * Backs the Guess screen: marks the atom STARTED, loads the previous atom's drawing for
 * replay, holds the in-progress guess text, and commits the guess once the replay has
 * finished and the field is non-blank.
 */
class GuessViewModel(
    private val container: AppContainer,
    private val gameId: String,
    threadIndex: Int,
    atomIndex: Int,
) : ViewModel() {

    private val address = AtomAddress(threadIndex, atomIndex)
    private val uid: String get() = container.auth.current()?.uid ?: ""

    /** Replay events for the drawing this guess responds to; empty until [prepare] resolves. */
    var events: List<DrawingEvent> = emptyList()
        private set

    /** Current text in the guess field. */
    var guess by mutableStateOf("")

    /** Set true by ReplayCanvas once the playback completes; gates submission. */
    var drawingFinished by mutableStateOf(false)

    private val _ready = MutableStateFlow(false)
    /** True after [prepare] has loaded the replay events. */
    val ready: StateFlow<Boolean> = _ready.asStateFlow()

    private val _submitted = MutableStateFlow(false)
    /** True once the guess has been committed via finishGuess. */
    val submitted: StateFlow<Boolean> = _submitted.asStateFlow()

    /** Submission is allowed once there's a non-blank guess and the replay has played out. */
    val canSubmit: Boolean
        get() = guess.isNotBlank() && drawingFinished && !_submitted.value

    private var prepared = false

    /** Run once: mark the atom STARTED and load the drawing to replay. */
    fun prepare() {
        if (prepared) return
        prepared = true
        viewModelScope.launch {
            container.games.setAtomState(gameId, address, GameAtomState.STARTED)
            val game = container.games.observeGame(gameId).first()
            val drawingKey = GameNavigation.guessDrawingRef(game.threads, address)
            events = drawingKey?.let { container.drawings.loadEvents(it) } ?: emptyList()
            _ready.value = true
        }
    }

    /** Commit the guess; no-op unless [canSubmit]. */
    fun submit() {
        if (!canSubmit) return
        val value = guess.trim()
        viewModelScope.launch {
            container.games.finishGuess(gameId, address, value, uid)
            _submitted.value = true
        }
    }
}

File: ios/Epyc/Features/Guess/GuessViewModel.swift
import Foundation
import SwiftUI
import EpycCore

/// State + side effects for the Guess screen (port of the legacy guess page).
///
/// Responsibilities:
/// - Mark the atom `.started` on appear.
/// - Load the drawing's recorded events so `ReplayCanvasView` can replay them.
/// - Gate submission on a non-empty guess AND the replay having finished.
/// - Persist the guess via `GameRepository.finishGuess`.
///
/// Shared services live in the SwiftUI environment, so the View passes them into
/// the VM's effect methods rather than the VM owning copies.
@MainActor
final class GuessViewModel: ObservableObject {

    // MARK: Inputs
    let gameId: String
    let drawingKey: String
    let address: AtomAddress

    // MARK: Published state
    /// Recorded events to replay; populated in `load`.
    @Published private(set) var events: [DrawingEvent] = []
    /// True once the events have been fetched (so the replay only starts with real data).
    @Published private(set) var loaded = false
    /// True once the replay animation has completed.
    @Published var drawingFinished = false
    /// The user's text guess.
    @Published var guess = ""
    /// Set once the guess has been submitted to avoid double-submission.
    @Published private(set) var didSubmit = false

    @Published var showLeaveConfirmation = false

    private var didLoad = false

    /// Submit is allowed only when there is a non-empty (trimmed) guess and the
    /// drawing replay has finished.
    var canSubmit: Bool {
        !guess.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
            && drawingFinished
            && !didSubmit
    }

    init(gameId: String, threadIndex: Int, atomIndex: Int, drawingKey: String) {
        self.gameId = gameId
        self.drawingKey = drawingKey
        self.address = AtomAddress(threadIndex: threadIndex, atomIndex: atomIndex)
    }

    // MARK: - Lifecycle

    /// Mark the atom started and load the drawing's events. Runs exactly once.
    func load(di: DIContainer) async {
        guard !didLoad else { return }
        didLoad = true
        defer { loaded = true }
        do {
            try await di.games.setAtomState(gameId: gameId, address: address, state: .started)
            events = try await di.drawings.loadEvents(drawingId: drawingKey)
        } catch {
            events = []
        }
    }

    // MARK: - Submit

    func submit(di: DIContainer, auth: AuthService) {
        guard canSubmit else { return }
        let trimmed = guess.trimmingCharacters(in: .whitespacesAndNewlines)
        let uid = auth.user?.uid ?? ""
        didSubmit = true
        Task { [weak self] in
            guard let self else { return }
            do {
                try await di.games.finishGuess(gameId: self.gameId,
                                               address: self.address,
                                               guess: trimmed,
                                               authorUid: uid)
            } catch {
                // Re-enable submission on failure.
                self.didSubmit = false
            }
        }
    }

    // MARK: - Navigation

    func confirmLeave() { showLeaveConfirmation = true }
}
